### PR TITLE
pi: Add ttl support

### PR DESCRIPTION
**Pull request by masco-enovance**

### CODE REVIEW

**Comment by masco-enovance** [comment]
run skydive-k8s-tests

**Comment by masco-enovance** [comment]
run skydive-python-tests

File: cmd/client/packet_injector.go [hunk 1]
```diff
@@ -198,6 +200,7 @@ func addInjectPacketFlags(cmd *cobra.Command) {
 	cmd.Flags().Int64VarP(&incrementPayload, "incrementPayload", "", 0, "increase payload for each packet")
 	cmd.Flags().Int64VarP(&count, "count", "", 1, "number of packets to be generated")
 	cmd.Flags().Int64VarP(&interval, "interval", "", 1000, "wait interval milliseconds between sending each packet")
+	cmd.Flags().Int64VarP(&ttl, "ttl", "", 0, "ttl in seconds")
```
Comment (safchain) [inline cmd/client/packet_injector.go]: TTL is not in second, it is the number of Hop the packet is allowed to pass through.

File: contrib/python/api/skydive/packet_injector.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ class PacketInjection(object):
     def __init__(self, uuid="", src="", dst="", srcip="", dstip="", srcmac="",
                  dstmac="", srcport="", dstport="", type="", payload="",
                  trackingid="", icmpid="", count="", interval="",
-                 increment="", starttime=""):
+                 increment="", starttime="", ttl=""):
```
Comment (safchain) [inline contrib/python/api/skydive/packet_injector.py]: this should be 255 by default. This makes me think that count, interval, increment should be numbers

File: contrib/python/api/skydive/rest/client.py [hunk 3]
```diff
@@ -220,7 +220,7 @@ def injection_create(self, src_query="", dst_query="", type="icmp4",
                          payload="", interval=1000, src_ip="",
                          dst_ip="", src_mac="", dst_mac="",
                          count=1, icmp_id=0, src_port=0, dst_port=0,
-                         increment=False):
+                         increment=False, ttl=0):
```
Comment (safchain) [inline contrib/python/api/skydive/rest/client.py]: 255 by default

File: contrib/python/api/skydive/packet_injector.py (hunk 2 above)
Comment (masco-enovance) [reply contrib/python/api/skydive/packet_injector.py]: modified all the fields

File: cmd/client/packet_injector.go (hunk 1 above)
Comment (masco-enovance) [reply cmd/client/packet_injector.go]: oops :-) modified.

File: contrib/python/api/skydive/rest/client.py (hunk 3 above)
Comment (masco-enovance) [reply contrib/python/api/skydive/rest/client.py]: done!

**Comment by masco-enovance** [comment]
run skydive-python-tests

File: api/types/types.go [hunk 4]
```diff
@@ -161,6 +161,7 @@ type PacketInjection struct {
 	IncrementPayload int64 `yaml:"IncrementPayload"`
 	StartTime        time.Time
 	Pcap             []byte `yaml:"Pcap"`
+	TTL              int64  `yaml:"TTL"`
```
Comment (hunchback) [inline api/types/types.go]: TTL should be a `unint8` (0-255)

https://en.wikipedia.org/wiki/Time_to_live

```
The maximum TTL value is 255, the maximum value of a single octet
```

File: cmd/client/packet_injector.go [hunk 5]
```diff
@@ -198,6 +200,7 @@ func addInjectPacketFlags(cmd *cobra.Command) {
 	cmd.Flags().Int64VarP(&incrementPayload, "incrementPayload", "", 0, "increase payload for each packet")
 	cmd.Flags().Int64VarP(&count, "count", "", 1, "number of packets to be generated")
 	cmd.Flags().Int64VarP(&interval, "interval", "", 1000, "wait interval milliseconds between sending each packet")
+	cmd.Flags().Int64VarP(&ttl, "ttl", "", 255, "time-to-live")
```
Comment (hunchback) [inline cmd/client/packet_injector.go]: s/255/64/

https://en.wikipedia.org/wiki/Time_to_live

```
A recommended initial value is 64
```

File: packetinjector/forge.go [hunk 6]
```diff
@@ -43,7 +43,7 @@ type ForgedPacketGenerator struct {
 	srcIP, dstIP   net.IP
 }
 
-func forgePacket(packetType string, layerType gopacket.LayerType, srcMAC, dstMAC net.HardwareAddr, srcIP, dstIP net.IP, srcPort, dstPort int64, ID int64, data string) ([]byte, gopacket.Packet, error) {
+func forgePacket(packetType string, layerType gopacket.LayerType, srcMAC, dstMAC net.HardwareAddr, srcIP, dstIP net.IP, srcPort, dstPort, TTL int64, ID int64, data string) ([]byte, gopacket.Packet, error) {
```
Comment (hunchback) [inline packetinjector/forge.go]: nit: I would rather add `TTL` before the `srcIP, dstIP ..` in the relative position it has in the IPv4 header:

https://en.wikipedia.org/wiki/IPv4

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-orientdb

File: api/types/types.go (hunk 4 above)
Comment (safchain) [reply api/types/types.go]: I agree this will avoid uint8(ttl) later

File: contrib/python/api/skydive/packet_injector.py [hunk 7]
```diff
@@ -22,8 +22,8 @@ class PacketInjection(object):
 
     def __init__(self, uuid="", src="", dst="", srcip="", dstip="", srcmac="",
                  dstmac="", srcport="", dstport="", type="", payload="",
-                 trackingid="", icmpid="", count="", interval="",
-                 increment="", starttime=""):
+                 trackingid="", icmpid="", count=1, interval=0,
+                 increment=0, starttime="", ttl=255):
```
Comment (safchain) [inline contrib/python/api/skydive/packet_injector.py]: See the comment of @hunchback, better to use 64

File: contrib/python/api/skydive/rest/client.py [hunk 8]
```diff
@@ -220,7 +220,7 @@ def injection_create(self, src_query="", dst_query="", type="icmp4",
                          payload="", interval=1000, src_ip="",
                          dst_ip="", src_mac="", dst_mac="",
                          count=1, icmp_id=0, src_port=0, dst_port=0,
-                         increment=False):
+                         increment=False, ttl=255):
```
Comment (safchain) [inline contrib/python/api/skydive/rest/client.py]: same here

File: contrib/python/api/skydive/packet_injector.py (hunk 7 above)
Comment (safchain) [reply contrib/python/api/skydive/packet_injector.py]: Sorry @masco-enovance my bad `Increment` should be set to `False` per the API definition

**Comment by masco-enovance** [comment]
run skydive-python-tests

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-orientdb

File: api/types/types.go [hunk 9]
```diff
@@ -169,6 +170,9 @@ func (pi *PacketInjection) Validate() error {
 	if _, ok := allowedTypes[pi.Type]; !ok {
 		return errors.New("given type is not supported")
 	}
+	if pi.TTL < 0 || pi.TTL > 255 {
```
Comment (safchain) [inline api/types/types.go]: this test is not useful as the type is uint8

File: packetinjector/forge.go [hunk 10]
```diff
@@ -61,7 +61,7 @@ func forgePacket(packetType string, layerType gopacket.LayerType, srcMAC, dstMAC
 
 	switch packetType {
 	case "icmp4":
-		ipLayer := &layers.IPv4{Version: 4, SrcIP: srcIP, DstIP: dstIP, Protocol: layers.IPProtocolICMPv4, TTL: 64}
+		ipLayer := &layers.IPv4{Version: 4, SrcIP: srcIP, DstIP: dstIP, Protocol: layers.IPProtocolICMPv4, TTL: uint8(TTL)}
```
Comment (safchain) [inline packetinjector/forge.go]: uint8(TTL) is not need anymore, no ?

File: api/types/types.go (hunk 4 above)
Comment (masco-enovance) [reply api/types/types.go]: done!

File: api/types/types.go (hunk 9 above)
Comment (masco-enovance) [reply api/types/types.go]: removed

File: packetinjector/forge.go (hunk 10 above)
Comment (masco-enovance) [reply packetinjector/forge.go]: yes, forgot to remove. now removed

**Comment by masco-enovance** [comment]
run skydive-python-tests

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-orientdb

**Pull request merged**